lib/lua/lua_init.lua

```diff
@@ -65,3 +65,59 @@ config.register_key = function(key)
         error("keybinding action: expected string or a function"..use_key, 2)
     end
 end
+-- Register callback to execute on restart
+way_cooler_table.on_restart = function(callback)
+    assert(callback, "missing callback")
+    assert(type(callback) == 'function', "callback: expected function")
+    __rust.restart_callback = callback
+end
+-- Register a function to execute on terminate
+way_cooler_table.on_terminate = function(callback)
+    assert(callback, "missing callback")
+    assert(type(callback) == 'function', "callback: expected function")
+    __rust.on_terminate = callback
+end
+-- This could technically be called by clients if they want, it should be more hidden.
+way_cooler_table.handle_termination = function()
```

Comment:
Clients can't kill the wm with this function, they can only call the Lua callback that does cleanup - the callback itself should be safe to call, at most it would just be annoying for any Lua code the user set up, and would probably cause the Lua thread to start erroring if other functions were called.

For example, if a Lua init file were to open a file, register a keybinding to read/write to the file, and in the `terminate`/`restart` close the file, then maliciously calling `way_cooler.handle_restart()` would close their open file and each keypress would probably error. 

The biggest concern we have with Lua right now is the various `dmenu-lua-do-things` commands - those still allow any program to register any Lua code to handle any keybinding. I don't think having this exposed is a good idea at all, but we should wait for a later PR because even if we did sandbox this a little we have bigger concerns.